include/libtorrent/aux_/websocket_tracker_connection.hpp

```diff
@@ -0,0 +1,130 @@
+/*
+
+Copyright (c) 2019 Paul-Louis Ageneau
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_WEBSOCKET_TRACKER_CONNECTION_HPP_INCLUDED
+#define TORRENT_WEBSOCKET_TRACKER_CONNECTION_HPP_INCLUDED
+
+#include "libtorrent/config.hpp"
+
+#if TORRENT_USE_RTC
+
+#include "libtorrent/aux_/rtc_signaling.hpp" // for rtc_offer and rtc_answer
+#include "libtorrent/aux_/websocket_stream.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/io_context.hpp"
+#include "libtorrent/peer_id.hpp"
+#include "libtorrent/aux_/resolver_interface.hpp"
+#include "libtorrent/tracker_manager.hpp" // for tracker_connection
+#include "libtorrent/ssl.hpp"
+
+#include <boost/beast/core/flat_buffer.hpp>
+
+#include <map>
+#include <memory>
+#include <queue>
+#include <tuple>
+#include <variant>
+
+namespace libtorrent {
+namespace aux {
+
+struct tracker_answer
+{
+	sha1_hash info_hash;
+	peer_id pid;
+	aux::rtc_answer answer;
+};
+
+struct TORRENT_EXTRA_EXPORT websocket_tracker_connection : tracker_connection
```

Comment:
at a cursory reading of this, it almost looks like `websocket_connection` could be merged with this class, and there would be less code. Maybe it's just a sign of me not fully appreciating the facilities provided by `websocket_connection` though. It primarily provides the abstraction of supporting both SSL and plain TCP, right?

Reply:
Do you mean with `websocket_stream`? It could be indeed, I just prefer to deal with purely websocket stuff like the states of the connection process and like you said abstracting SSL and non-SSL websockets.

Reply:
yes, I mean `websocket_stream`